src/headless/converse-muc.js

```diff
@@ -941,74 +942,40 @@ converse.plugins.add('converse-muc', {
-                } else if (nicknames.includes(longest_match)) {
-                    // TODO: show a warning to the user that the person is not currently in the chat
-                    uri = `${this.get('jid')}/${longest_match}`;
-                } else {
-                    return;
-                }
-                const obj = {
-                    'begin': index,
-                    'end': index + longest_match.length,
-                    'value': longest_match,
-                    'type': 'mention',
-                    'uri': encodeURI(`xmpp:${uri}`)
-                };
-                return obj;
+            getOccupantByJID (jid) {
+                return u.isValidJID(jid) && this.occupants.findOccupant({ jid });
             },
 
-            extractReference (text, index) {
-                for (let i=index; i<text.length; i++) {
-                    if (text[i] === '@' && (i === 0 || text[i - 1] === ' ')) {
-                        const match = text.slice(i+1),
-                              ref = this.getReferenceForMention(match, i);
-                        if (ref) {
-                            return [text.slice(0, i) + match, ref, i]
-                        }
-                    }
-                }
-                return;
+            getOccupantByNickname (nick) {
+                return this.occupants.findOccupant({ nick });
             },
 
-            parseTextForReferences (text) {
-                const refs = [];
-                let index = 0;
-                while (index < (text || '').length) {
-                    const result = this.extractReference(text, index);
-                    if (result) {
-                        text = result[0]; // @ gets filtered out
-                        refs.push(result[1]);
-                        index = result[2];
-                    } else {
-                        break;
-                    }
-                }
-                return [text, refs];
+            parseTextForReferences (original_message) {
+                const dirty_mentions = [...original_message.matchAll(p.mention_regex)];
+                const mentions = dirty_mentions.map(p.mapMentionsInMatches);
+                if (!mentions.length) return [original_message, []];
+
+                const known_nicknames = this.getAllKnownNicknames();
+                const indexes = dirty_mentions.map(p.mapMatchesToBeginIndexes(/[\s]+/));
+                const make_uri_from_reference = p.makeUriFromReference(
+                    this.getOccupant.bind(this),
+                    this.get('jid')
+                )
+
+                const references = mentions
+                    .map(p.mapMentionsToTempReferences)
+                    .reduce(p.reduceReferencesWithNicknames(known_nicknames), [])
+                    .map(p.mapAddUriToReferences(make_uri_from_reference))
+                    .map(p.mapAddCoordsToReferences(indexes))
+                    .map(p.mapCleanReferences);
```

Comment:
Instead of mapping over the array three times after `reduce` has been called, you can use the `flow` method from lodash to map only once.

For example (untested):
```
const finalizeReference = flow([
    p.mapAddUriToReferences(make_uri_from_reference),
    p.mapAddCoordsToReferences(indexes),
    p.mapCleanReferences
]);

const reducer = p.reduceReferencesWithNicknames(known_nicknames);

const references = mentions
    .map(p.mapMentionsToTempReferences)
    .reduce(reducer, [])
    .map(finalizeReference);
```

This way, we avoid 2 additional iterations through the array and we also don't repeatedly invoke `reduceReferencesWithNicknames`, `mapAddUriToReferences` and `mapAddCoordsToReferences` (some of which in turn call `map` as well).

Reply:
Please review the new code.
This is still a WIP since there are failing tests.